lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -1401,6 +1402,188 @@ static void maybeDiagnoseClassWithoutInitializers(ClassDecl *classDecl) {
   diagnoseClassWithoutInitializers(classDecl);
 }
 
+static bool isModuleQualified(TypeRepr *repr, ModuleDecl *module) {
+  auto extIdent = dyn_cast_or_null<CompoundIdentTypeRepr>(repr);
+  if (!extIdent) {
+    return false;
+  }
+  
+  auto components = extIdent->getComponents();
+  if (components.size() == 1) {
+    return false;
+  }
+  
+  return components.front()->getNameRef().isSimpleName(module->getName());
+}
+
+static void diagnoseRetroactiveConformances(ExtensionDecl *ext) {
+  ModuleDecl *module = ext->getParentModule();
+
+  // We only want to warn about this if a module has library evolution enabled,
+  // which we assume means it's a library meant to be distributed to clients.
+  if (!module->isResilient()) {
+    return;
+  }
+
+  // Disable this for the Foundation module because of the interplay with
+  // _ObjectiveCBridgeable.
+  if (module->isFoundationModule()) {
+    return;
+  }
+
+  // Don't warn for this if we see it in module interfaces (that usually
+  // indicates the interface was built with
+  // -module-interface-preserve-types-as-written)
+  if (ext->getParentSourceFile()->Kind == SourceFileKind::Interface) {
+    return;
+  }
+
+  Type extendedType = ext->getExtendedType();
+  NominalTypeDecl *extendedNominalDecl = ext->getExtendedNominal();
+  if (!extendedNominalDecl) {
+    return;
+  }
+
+  // If it's a type in the same module, conformances aren't retroactive.
+  ModuleDecl *extTypeModule = extendedNominalDecl->getParentModule();
+  if (extTypeModule == module) {
+    return;
+  }
+
+  // If we're compiling an overlay and the type comes from the clang module,
+  // conformances aren't retroactive.
+  if (auto clangModule = extTypeModule->findUnderlyingClangModule()) {
+    if (clangModule->getTopLevelModuleName() == module->getName().str()) {
```

Comment:
I wonder if we even need to go this far. I think all you need to do is something like

```
static bool isClangOverlayOf(ModuleDecl *potentialOverlay,
                             ModuleDecl *potentialUnderlying) {
  return !potentialOverlay->isNonSwiftModule() &&
         potentialUnderlying->isNonSwiftModule() &&
         potentialOverlay->getName() == potentialUnderlying->getName();
}
```

If you promoted this to API on `ModuleDecl` that'd be swell.